### DISCUSSION TITLE
For the knowledge graph created, which service is it stored in?

### DISCUSSION THREAD

**Discussion by chakpongchung** [microsoft/graphrag#328] [General]
I don't seem to see a graph store used in the project. For the topology part of the knowledge graph, which service is it stored in so people can query against?

**Comment by jgbradley1** [comment]
The knowledge graph is saved off as a collection of parquet files locally.

The output of an indexing run is defined by the `storage` section of the configuration file. See [here](https://microsoft.github.io/graphrag/posts/config/json_yaml/) for a description of all configuration values. For example, if `storage.type=file`, a directory with the name `storage.base_dir` (default value=`output`) will contain all the parquet files from the indexing run.

Other storage options are available (i.e. Azure Storage) which you can configure graphrag to use if you wish.

**Reply by chakpongchung** [reply]
Thanks. Does there exist a query execution engine to allow users to query/traverse the graph through some DSLs?  Or is prompting  through natural language the only way to interact with this project?

**Reply by jgbradley1** [reply]
The graphrag package contains both an indexing engine and a query engine. See [here](https://microsoft.github.io/graphrag/posts/query/3-cli/) for how to use the query engine in a no-code approach via the command line. For a programmatic approach, we have a couple of [notebooks](https://microsoft.github.io/graphrag/posts/query/notebooks/overview/) that demonstrate how to execute a query over the knowledge graph.

**Reply by hclgit2rahul** [reply]
> The knowledge graph is saved off as a collection of parquet files locally.
> 
> The output of an indexing run is defined by the `storage` section of the configuration file. See [here](https://microsoft.github.io/graphrag/posts/config/json_yaml/) for a description of all configuration values. For example, if `storage.type=file`, a directory with the name `storage.base_dir` (default value=`output`) will contain all the parquet files from the indexing run.
> 
> Other storage options are available (i.e. Azure Storage) which you can configure graphrag to use if you wish.

@jgbradley1 : Thanks for clarifying the query on storage location. I have a follow-up question, based on mentioned configuration, I am able to store parquet files on azure blob location with similar folder structure as it was in local. However, the same configuration is not picked for graphrag.query, it continue to looks for local parquet files. Is there a switch or configuration which instruct it look it in blob storage ?

**Reply by jesusfbes** [reply]
I have the same question of @hclgit2rahul. @jgbradley1 would you mind to help us with this?

**Reply by hclgit2rahul** [reply]
@jesusfbes @jgbradley1 - I spent some more time, it seems that programmatically we can read data frames from paraquet file stored in azure blob and the use them for search and LLM prompts to get the outcome. However, using CLI - I do not see an option.

**Reply by AR-javis** [reply]
Is it possible to use it with AWS S3? Or is Azure the only one supported? Thanks in advance

**Reply by jgbradley1** [reply]
@hclgit2rahul @jesusfbes you are correct that the query CLI only worked on local parquet files. We have made several major changes internal to the package over the last several months to get indexing and query more in sync with each other and provide support for the same functionality (i.e. utilizing azure storage as a storage backend).

If you haven't tried the latest release of graphrag, please give it a try and open an issue if the problem still exist (I recall this particular issue did get addressed though).

**Comment by reachrkr** [comment]
How can we store in other formats used by popular Graph DB's like neo4j

**Reply by jgbradley1** [reply]
The quickest approach would be to write a small script that transforms the parquet file data to a format (i.e. CSV) that is acceptable by the graph DB that you wish to use. After running graphrag indexing, you would then want to run this data transformation script over the parquet files.

**Comment by jli113** [comment]
Already have the graph file and loaded it in neo4j, can we use graphs in GraphDB directly instead of the indexing process?

**Comment by esbenclarlund** [comment]
How to use the vector db in Azure. Preferable I would like to configure graphRAG to write to Azure ai search directly but I do not find that option but to get it into a blob will do.
I already got it to write the .parquet files to a blob on generation but do not know how to utilize what by default is written to lancedb on my local disk.
Any input will be highly appreciated, thanks.

**Comment by rockerritesh** [comment]
@jgbradley1 Can you help me to connect min.io based lanced db. I have added the uri to connect particular buckets but doesnt seem to connect at server, but instead folder was connected on local. How to fix this.
![image](https://github.com/user-attachments/assets/37cbbbf8-9147-4bac-a581-5581a291d50b)
